### PR TITLE
[Metrics] Fairseq BLEU Re-Implementation

### PR DESCRIPTION
**Pull request by klshuster**
**Patch description**
Prior experiments for e.g. [dodecadialogue](https://parl.ai/projects/dodecadialogue/) have utilized fairseq's tokenized bleu scores for measuring BLEU. From inspection of the [fairseq generation script](https://github.com/pytorch/fairseq/blob/master/fairseq_cli/generate.py), it becomes apparent that they calculate BLEU on a macro level, considering all predictions and references together. Our original implementation did not take into account the `brevity` score, which is a total across all predictions/references, and thus our scores were misaligned.

I've thus re-implemented fairseq's bleu scoring with ParlAI's metrics system. I have verified that the computed BLEU score for e.g. the command referenced in #3473 yields nearly the same BLEU score as reported in the paper.

**Testing steps**
Added tests to CI:

```
$ pytest -k TestFairseqBleuMetric -v
====test session starts ====
...
collected 477 items / 476 deselected / 1 selected

test_metrics.py::TestFairseqBleuMetric::test_scorer PASSED                                                                                                                     [100%]

====slowest 10 durations ====
0.12s call     tests/test_metrics.py::TestFairseqBleuMetric::test_scorer

(2 durations < 0.005s hidden.  Use -vv to show these durations.)
====1 passed, 476 deselected, 3 warnings in 5.13s ====
```

Additionally:

```
$ parlai eval_model -mf zoo:dodecadialogue/empathetic_dialogues_ft/model -t empathetic_dialogues --skip-generation false -dt test --metrics ppl,bleu --inference beam --beam-size 10 --beam-min-length 5 --beam-block-ngram 3 --beam-context-block-ngram -1 --compute-tokenized-bleu true -bs 32
.
.
.
 fairseq_bleu1  fairseq_bleu2  fairseq_bleu3  fairseq_bleu4 
  15.61         8.018           5.159            3.485
```

### CODE REVIEW

File: parlai/core/metrics.py [hunk 1]
```diff
@@ -458,7 +459,103 @@ def compute(guess: str, answers: List[str], k: int = 4) -> Optional[BleuMetric]:
...
+        self.stat.count1 += other.stat.count1
+        self.stat.count2 += other.stat.count2
+        self.stat.count3 += other.stat.count3
+        self.stat.count4 += other.stat.count4
+        self.stat.predlen += other.stat.predlen
+        self.stat.reflen += other.stat.reflen
+        return self
+
+    def precision(self):
+        def ratio(a, b):
```
Comment (stephenroller) [inline parlai/core/metrics.py:535]: Please no closures

File: parlai/core/metrics.py [hunk 2]
```diff
@@ -458,7 +459,103 @@ def compute(guess: str, answers: List[str], k: int = 4) -> Optional[BleuMetric]:
...
+        self.stat.match4 += other.stat.match4
+        self.stat.count1 += other.stat.count1
+        self.stat.count2 += other.stat.count2
+        self.stat.count3 += other.stat.count3
+        self.stat.count4 += other.stat.count4
+        self.stat.predlen += other.stat.predlen
+        self.stat.reflen += other.stat.reflen
+        return self
+
+    def precision(self):
```
Comment (stephenroller) [inline parlai/core/metrics.py:534]: Make _precision?

File: tests/test_metrics.py [hunk 3]
```diff
@@ -439,5 +442,61 @@ def test_intra_distinct(self):
...
+class TestFairseqBleuMetric(unittest.TestCase):
+    """
+    We're just going to compare that scores from Fairseq's Bleu scorer are the same as
+    our scorer.
+    """
+
+    def test_scorer(self):
+        import random
+
+        random.seed(42)
```
Comment (stephenroller) [inline tests/test_metrics.py:455]: Is this needed?

File: tests/test_metrics.py [hunk 4]
```diff
@@ -439,5 +442,61 @@ def test_intra_distinct(self):
...
+                BleuConfig(pad=pad_idx, eos=eos_idx, unk=unk_idx)
+            )
+        except ImportError:
+            # Bleuconfig is a recent version of fairseq
+            fairseq_metrics: Scorer = Scorer(pad_idx, eos_idx, unk_idx)
+
+        parlai_metrics: Dict[int, FairseqBleuMetric] = {k: [] for k in range(1, 5)}
+
+        for _ in range(num_ex):
+            guess = torch.LongTensor(random.sample(range(vocab_length), ex_length))
```
Comment (stephenroller) [inline tests/test_metrics.py:477]: Would it be better to just make some deterministic, fixed exs?

**Review by stephenroller — APPROVED** [review]
Great sleuthing

File: parlai/core/metrics.py (hunk 1 above)
Comment (klshuster) [reply parlai/core/metrics.py:535]: sure

File: tests/test_metrics.py (hunk 3 above)
Comment (klshuster) [reply tests/test_metrics.py:455]: probably not

File: tests/test_metrics.py (hunk 4 above)
Comment (klshuster) [reply tests/test_metrics.py:477]: would you prefer that? I wanted to make sure enough examples were given that the brevity comes into play, which might be difficult with deterministic/fixed exs

Comment (klshuster) [reply tests/test_metrics.py:475]: i'm going to keep as is and we can revisit if it becomes an issue

**Pull request merged**